### PR TITLE
Add slice test support for SDC

### PR DESCRIPTION
**Pull request by eddumelendez**
See gh-8607

### CODE REVIEW

File: spring-boot-test-autoconfigure/src/test/java/org/springframework/boot/test/autoconfigure/data/cassandra/DataCassandraTestReactiveIntegrationTests.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Sample tests for {@link DataCassandraTest} using reactive repositories.
+ *
+ * @author Eddú Meléndez
+ */
+@RunWith(SpringRunner.class)
+@DataCassandraTest
+@TestPropertySource(properties = {"spring.data.cassandra.schemaAction=recreate_drop_unused",
```
Comment (snicoll) [inline spring-boot-test-autoconfigure/src/test/java/org/springframework/boot/test/autoconfigure/data/cassandra/DataCassandraTestReactiveIntegrationTests.java]: Not sure I like the idea of having to define those. IMO, if we have to then users of the annotation will. And if they do, then we probably need an OOB property for this with a sensible value

File: spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/cassandra/DataCassandraTest.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ * Annotation that can be used in combination with {@code @RunWith(SpringRunner.class)}
+ * for a typical Cassandra test. Can be used when a test focuses <strong>only</strong> on
+ * Cassandra components.
+ * <p>
+ * Using this annotation will disable full auto-configuration and instead apply only
+ * configuration relevant to Cassandra tests.
+ * <p>
+ * By default, tests annotated with {@code @DataCassandraTest} will use an embedded in-memory
+ * Cassandra process (if available). They will also be transactional with the usual
+ * test-related semantics (i.e. rollback by default).
```
Comment (snicoll) [inline spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/cassandra/DataCassandraTest.java]: Is the transactional part really intended? Is it supported at all? In any case, there's no `@Transactional` on this annotation.

File: spring-boot-test-autoconfigure/src/test/java/org/springframework/boot/test/autoconfigure/data/cassandra/DataCassandraTestIntegrationTests.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	public void didNotInjectExampleService() {
+		this.thrown.expect(NoSuchBeanDefinitionException.class);
+		this.applicationContext.getBean(ExampleService.class);
+	}
+
+	private void createTestKeyspaceIfNotExists() {
+		Session session = this.server.getCluster().connect();
+		try {
+			session.execute("CREATE KEYSPACE IF NOT EXISTS boot_test"
+					+ "  WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 1 };");
```
Comment (snicoll) [inline spring-boot-test-autoconfigure/src/test/java/org/springframework/boot/test/autoconfigure/data/cassandra/DataCassandraTestIntegrationTests.java]: Curious. Isn't it something we could do as well? If a user of embedded cassandra has to do that manually, then why would we have a dedicated annotation for it?

**Comment by eddumelendez** [comment]
@snicoll what about adding `@CassandraTestPropertySource` or something like that which allows to set the `schemaAction` and `keyspaceName`. WDYT?

**Comment by snicoll** [comment]
That is one part but it only has an added value when sensible defaults are used. But maybe we don't have sensible defaults. I am not using cassandra so I don't know which strategy you'd use.

**Comment by eddumelendez** [comment]
in this case, I think a new attribute `script` in the new annotation can help. So, it will have high priority than `keyspaceName`, otherwise we can just create one by default using `keyspaceName`.

**Comment by snicoll** [comment]
No need for a new name, stuff you put in the annotation always have precedence over the other property source. I am afraid I am not sure I understand what would be the plan. If those two properties must be present, what can we do there for the user?

**Comment by snicoll** [comment]
@eddumelendez any idea? I am not using Cassandra enough to have an opinion but I think someone should before we proceed with this one.

**Comment by snicoll** [comment]
@eddumelendez any idea about this (or anyone watching and using Cassandra)?

**Comment by eddumelendez** [comment]
@snicoll sorry for the delay. I think that at least boot should provide the following parameters in `@DataCassandraTest(schemaAction= SchemaAction.RECREATE_DROP_UNUSED, keyspaceName = "boot_test")`, where `schemaAction` is optional but `keyspaceName` is required. IMHO, I think that the script we are using in the test to create the `keyspace` should be performed in cassandra previous to the test execution (i.e create a container with that keyspace and run the test against that container).

I am not expert in this subject but that would work.

**Comment by snicoll** [comment]
You can't make the keyspace mandatory if it has been specified in user configuration I am afraid. To be clear I am looking for at is:

* Do you need to be able to change the `keyspaceName` per test?
* If no `keyspaceName` is provided in user configuration, does inferring a default the right thing to do?
* Does overriding the `schemaAction` for tests a god thing (considering we can or not infer a `keyspaceName`)

In the end, if we can add `@DataCassandraTest` and have something running on a "test" keyspace name with schema generation is the right thing to do, we should do that. And if changing the keyspace per test doesn't make sense, we shouldn't define a parameter for it IMO.

perhaps @christophstrobl or @mp911de have an opinion?

File: spring-boot-test-autoconfigure/src/test/java/org/springframework/boot/test/autoconfigure/data/cassandra/DataCassandraTestReactiveIntegrationTests.java (hunk 1 above)
Comment (mp911de) [inline spring-boot-test-autoconfigure/src/test/java/org/springframework/boot/test/autoconfigure/data/cassandra/DataCassandraTestReactiveIntegrationTests.java]: `recreate_drop_unused` removes the content of the entire keyspace (database). Configuration makes sense but it should stay Boot's default

**Comment by mp911de** [comment]
`keyspaceName` is required to create a sensible `Session` object. Otherwise, the session object connects to the `system` keyspace which isn't usable for user operations. 

Regarding `schemaAction` it depends on what the user strategy, whether they are using Spring Data's schema generation or the use other Cassandra provisioning mechanisms.

Aligning these two properties with relational database test support would make the most sense (like `spring.jpa.generate-ddl`). What also plays into a convenient test support would be auto-execution of possibly existing `schema.cql` and `data.cql` files, but that's a different, not-yet-existing, feature (similar to `schema.sql` and `data.sql`).

**Comment by snicoll** [comment]
It doesn't look like we have a clear view of what  that slice should do so I am going to close this one. We can certainly reconsider if we have more input.

**Pull request closed without merging**